Add shorthand properties reference to public API

ybiquitous opened the pull request:
<!-- Each pull request must be associated with an open issue unless it's a documentation fix. If a corresponding issue does not exist, please create one so we can discuss the change first. -->

<!-- Please answer the following. We close pull requests that don't. -->

> Which issue, if any, is this issue related to?

- Closes #6151
- #6152

> Is there anything in the PR that needs further explanation?

This pull request is following the suggestion below by @jeddy3 in <https://github.com/stylelint/stylelint/issues/6151#issuecomment-1160488202>:

> - align and expose data as it's requested (starting with the shorthand data)

The type definition is generated by the following script:

```js
const shorthandData = require("./lib/reference/shorthandData")
console.log('export type ShorthandData = {')
for (const key of Object.keys(shorthandData).sort()) {
	console.log(`\t'${key}': Set<string>;`)
}
console.log('};')
```

ybiquitous commented:
[ask] The `shorthandData` name seems unclear to me. Is it acceptable, or is there a better name?

jeddy3 commented:
@ybiquitous Thank you. It's looking good.

> [ask] The shorthandData name seems unclear to me. Is it acceptable, or is there a better name?

Let's go with `subPropertiesOfLonghandProperties` as per https://github.com/stylelint/stylelint/issues/6151#issuecomment-1160488202.

Future property sets could include [`acceptCustomIdentsProperties`](https://github.com/stylelint/stylelint/blob/main/lib/reference/propertySets.js)  and these:

https://github.com/stylelint/stylelint/blob/bd8c7fd88172af93d8ae80b38256b5074ebee233/lib/reference/keywordSets.js#L411-L423

(Which are in the `keywordSets.js` at the moment...)

What file structure shall we use? Choices are:

- 1 file per set, e.g. `reference/subPropertiesOfLonghandProperties.js`
- 1 file per group, e.g. `reference/properties.js` (which would contain all the above)
- a single file for all the sets, e.g. `reference/reference.js`

I feel like either of the last two will be easiest to work with and maintain.

ybiquitous commented:
@jeddy3 Thanks for your helpful feedback! The idea of "1 file per group" seems so nice. 👍🏼 

I'll push commits into this pull request, following the ideas of `subPropertiesOfLonghandProperties` and "1 file per group".

jeddy3 commented:
> I'll push commits into this pull request, following the ideas of `subPropertiesOfLonghandProperties` and "1 file per group".

Oops, I got it back-to-front. From the spec:

> Some properties are shorthand properties, meaning that they allow authors to specify the values of several properties with a single property. A [**shorthand property**](https://www.w3.org/TR/css-cascade-4/#shorthand-property) sets all of its **longhand sub-properties**, exactly as if expanded in place.

Let's go with `longhandSubPropertiesOfShorthandProperties` to be accurate with the wording in the spec, even if it's a bit verbose.

ybiquitous commented:
I have one more suggestion: what if using `Map` instead of `Object`?

[This MDN doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map#objects_vs._maps) describes the differences between `Map` and `Object` well. I think `Map` could be safer and more stable than `Object`, especially as a public API, even if `Map` doesn't support JSON serialization natively. E.g.

```js
// Using Object
const longhandSubPropertiesOfShorthandProperties = {
  margin: new Set(['margin-top', 'margin-bottom', 'margin-left', 'margin-right']),
  // ...
};
longhandSubPropertiesOfShorthandProperties.margin; //=> Set(...)

// Using Map
const longhandSubPropertiesOfShorthandProperties = new Map([
  ['margin', new Set(['margin-top', 'margin-bottom', 'margin-left', 'margin-right'])],
  // ...
]);
longhandSubPropertiesOfShorthandProperties.get('margin'); //=> Set(...)
```

What do you think?

ybiquitous commented on lib/reference/properties.js:
```diff
@@ -1,8 +1,18 @@
-	padding: new Set(['padding-top', 'padding-bottom', 'padding-left', 'padding-right']),
+/** @type {import('stylelint').LonghandSubPropertiesOfShorthandProperties} */
+const longhandSubPropertiesOfShorthandProperties = {
+	/* eslint sort-keys: 'error' */
```
[note] I've sorted properties alphabetically for findability (the diff increases, though).

ybiquitous commented on lib/reference/properties.js:
```diff
@@ -91,6 +47,18 @@ module.exports = {
 		'border-block-start-color',
 	]),
+	'border-bottom': new Set([
+		// prettier-ignore
```
[note] To align always multi-line for readability.

jeddy3 commented:
> I have one more suggestion: what if using Map instead of Object?

Sounds good to me. 

All the other changes are looking great, by the way!

ybiquitous commented on types/stylelint/index.d.ts:
```diff
@@ -383,6 +383,45 @@ declare module 'stylelint' {
+			| 'text-decoration'
+			| 'text-emphasis'
+			| 'transition',
+			ReadonlySet<string>
```
[note] `ReadonlyMap` and `ReadonlySet` are built-in types.

ybiquitous commented on lib/rules/declaration-block-no-shorthand-property-overrides/index.js:
```diff
@@ -27,34 +27,36 @@ const rule = (primary) => {
 
 				for (const longhandProp of overrideables) {
-					if (!Object.prototype.hasOwnProperty.call(declarations, prefix + longhandProp)) {
+					const declaration = declarations.get(prefix + longhandProp);
```
[note] Refactor to use `Map` instead of `Object`; `hasOwnProperty` is no longer needed (safe access).

jeddy3 replied:
Much nicer.

ybiquitous commented on lib/rules/declaration-block-no-redundant-longhand-properties/index.js:
```diff
@@ -40,49 +40,47 @@ const rule = (primary, secondaryOptions) => {
+				shorthands.push(shorthand);
+				longhandToShorthands.set(longhand, shorthands);
+			}
+		}
```
[note] Refactor to use `Map` and `for...of` instead of `Object` and `reduce()` for readability.

ybiquitous commented:
@jeddy3 I've switched from `Object` to `Map` and `Set` (`ReadonlyMap` and `ReadonlySet` in TypeScript), and also refactored `declaration-block-*` rules using `Map`.

jeddy3 reviewed (approved):
Fantastic, thank you!

I think we have a good direction here for exposing our reference data.

jeddy3 commented:
> - Added: longhand sub-properties of shorthand properties reference data to public API ([#6168](https://github.com/stylelint/stylelint/pull/6168)).

ybiquitous commented:
@jeddy3 Thank you for all of your reviews. 😊 👍🏼
